Add files via upload

vda-777 opened the pull request.

vda-777 committed:
Add files via upload

vda-777 committed:
Delete App.tsx

vda-777 committed:
Delete AppHeader.jsx

vda-777 committed:
Update burger-ingredients.jsx

gennady-bars commented on src/components/App.jsx:
```diff
@@ -0,0 +1,58 @@
+            `This is an HTTP error: The status is ${response.status} description is ${response.statusText}`
+          );
+        }
+        let resJson = await response.json();
```
Если переменная объявлена через `let`, её значение должно быть напрямую перезаписано.
Если переменная **не переназначается**, то она объявляется через `const`.


Все переменные нужно объявлять по возможности через `const`.
В больших функциях бывает так, что Вы можете не заметить, если переменная переназначится, и это может оказаться причиной багов. Особенно в замыканиях. 
`const` дает Вам уверенность в том, что Вы всегда будете видеть одно и то же значение.

gennady-bars commented on src/components/App.jsx:
```diff
@@ -0,0 +1,58 @@
+        //else {setFilteredData(allData);}
+  
+        setError(null);
+      } catch(err) {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/components/App.jsx:
```diff
@@ -0,0 +1,58 @@
+      }  
+    }
+    getData()
+  }, [])
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars commented on src/components/App.jsx:
```diff
@@ -0,0 +1,58 @@
+    <>
+      <AppHeader />
+
+      <div className='context' style={{display: 'flex', justifyContent: 'center', maxHeight: '87vh', minHeight: '87vh'}}>
```
`style={{` используете 37 раз в коде проекта. Их все нужно убрать


Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.

gennady-bars replied:
тут `div` нужно заменить на тег `main`


По семантике нужно помещать весь главный контент страницы после хэдера в тег `main`.

gennady-bars commented on src/components/app-header/app-header.jsx:
```diff
@@ -0,0 +1,22 @@
+
+export default function AppHeader() {
+  return (
+    <nav className='navigationPanel' >
```
в `AppHeader` по семантике должен быть тег `header`

gennady-bars replied:
**Можно лучше**

Чтобы сделать сайт более доступным для пользователя, нужно стараться придерживаться **семантической верстки**.
Несколько полезных статей на эту тему:
•	[Что такое семантическая вёрстка и зачем она нужна](https://htmlacademy.ru/blog/html/semantics)
•	[Семантика для циников](https://www.youtube.com/watch?v=B5pomLr059U)

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,78 @@
+    let sumdata = data.reduce((sum, item) => sum + item.price, 0) + 200 + 200;
+    return (
+      <>
+      {showModal && createPortal(
```
`createPortal` должен быть только внутри `Modal`, чтобы не дублировать его под каждый попап

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,78 @@
+      <>
+      {showModal && createPortal(
+        <OrderDetails onClose={() => setShowModal(false)} numberOrder="034356"/>,
+        document.body
```
` document.body`

Для **портала** по заданию нужно сделать отдельный `<div id="modals"></div>` в файле **index.html** в папке `public`. Скрин http://joxi.ru/EA4YWBvSXl1R9r

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,78 @@
+BurgerConstructor.propTypes= {
+  allData: PropTypes.arrayOf(
+    PropTypes.shape({
+    _id: PropTypes.string,
```
тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+BurgerIngredients.propTypes= {
+  allData: PropTypes.arrayOf(
+    PropTypes.shape({
+    _id: PropTypes.string,
```
тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+  })),
+  filteredData: PropTypes.arrayOf(
+    PropTypes.shape({
+    _id: PropTypes.string,
```
тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+  return (
+    <>
+    {showModal && createPortal(
+      <IngredientsDetails onClose={() => setShowModal(false)} currentIngredient={currentItem}/>,
```
не стоит использовать компонент `Modal` внутри компонентов с деталями ингредиентов и заказа. Это контент, а не модальные окна. Это усложнит задачу в следующих спринтах, где будут еще **страницы** с этим контентом (не модальные окна). Нужно было сделать отдельные компоненты с деталями и передавать их в качестве `children` в `Modal`
```jsx

    <Modal
      title={'Детали ингредиента'}
      onClose={handleClose}
    >
      <IngredientDetails currentIngredient={currentIngredient} /> 
    </Modal>

```

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,94 @@
+                {props.filteredData && Array.isArray(props.filteredData) &&
+                    props.filteredData.map(({_id, name, type, proteins, fat, carbohydrates, calories, price, image, image_large}) => (                      
+                      /*<li key={_id} className="Ingredient mr-6 mb-8" style={{listStyleType: `none`, maxWidth:'50%', alignContent:'stretch'}} onClick={(event) => setShowModal(true)}>*/
+                        <li key={_id} className="Ingredient mr-6 mb-8" style={{listStyleType: `none`, maxWidth:'50%', alignContent:'stretch'}} onClick={(event) => {ingredientClick({
```
`key={_id}`


Отлично, что используете атрибут `key`, и отлично, что для него используете уникальный `_id`, а не индекс массива.

[Почему индекс не нужно использовать, можно прочитать тут]( https://medium.com/russian/%D0%B8%D0%BD%D0%B4%D0%B5%D0%BA%D1%81-%D0%B2-%D0%BA%D0%B0%D1%87%D0%B5%D1%81%D1%82%D0%B2%D0%B5-%D0%BA%D0%BB%D1%8E%D1%87%D0%B0-%D0%B0%D0%BD%D1%82%D0%B8-%D0%BF%D0%B0%D1%82%D1%82%D0%B5%D1%80%D0%BD-56bab8ada309)

gennady-bars commented on src/components/ingredients-details/ingredients-details.jsx:
```diff
@@ -0,0 +1,59 @@
+    //zIndex: '1'
+  };
+    return (
+      <div style={styleModal} >        
+        <span className='mt-10 ml-10 mr-10' style={{display:'flex', justifyContent:'space-between', width:'auto'}}>
+          <p className="text text_type_digits-medium">Детали ингредиента</p><CloseIcon type="primary" onClick={props.onClose}/>
+        </span>
```
По заданию должен быть компонент `Modal` для обертки модальных окон. В нем будет крестик, обработчик оверлея и `Escape`. Внутри него будет портал и `ModalOverlay` (это темное пространство вокруг попапа)

gennady-bars commented on src/components/ingredients-details/ingredients-details.jsx:
```diff
@@ -0,0 +1,59 @@
+IngredientsDetails.propTypes={
+  onClose: PropTypes.func,
+  currentIngredient: PropTypes.shape({
+    image: PropTypes.string,
```
тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз
* Отлично, что используете атрибут `key`, и отлично, что для него используете уникальный `_id`, а не индекс массива.




но есть некоторые недочеты: 

* По заданию под каждый компонент должен быть свой собственный модульный файл `css`. 
* Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.
* `App.css` нужно удалить. Все из него нужно распределить по файлам `css`, которые будут относиться к конкретным компонентам
* `position: absolute` не стоит использовать в `css` в проекте. Это нигде не нужно (кроме крестика в попапе). Иначе невозможно сделать хоть какой-то минимальный адаптивный дизайн сайта. Все делается на `flex` или `grid` обычно, чтобы позиционировать правильно (и центрировать)
* Попапы должны закрываться по оверлею и `Escape` еще
* У попапа должен быть оверлей по заданию
* Если переменная **не переназначается**, то она объявляется через `const`.
* `style={{` используете 37 раз в коде проекта. Их все нужно убрать
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.
* По семантике нужно помещать весь главный контент страницы после хэдера в тег `main`.
* `App.jsx` нужно переместить в папку `components`, сделать с маленькой буквы, так как по заданию нужно называть файлы в `kebab-case`, и там поместить в свою папку `app`
* `App.css` нужно переместить в папку к `app.jsx` и сделать модулем `app.module.css`, а в компоненте использовать модульные классы, а не строки 
* в `AppHeader` по семантике должен быть тег `header`
* `createPortal` должен быть только внутри `Modal`, чтобы не дублировать его под каждый попап. В этом суть `Modal`
* Для **портала** по заданию нужно сделать отдельный `<div id="modals"></div>` в файле **index.html** в папке `public`. Скрин http://joxi.ru/EA4YWBvSXl1R9r 
* тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно
* не стоит использовать компонент `Modal` внутри компонентов с деталями ингредиентов и заказа. Это контент, а не модальные окна. Это усложнит задачу в следующих спринтах, где будут еще **страницы** с этим контентом (не модальные окна). Нужно было сделать отдельные компоненты с деталями и передавать их в качестве `children` в `Modal`
* По заданию должен быть компонент `Modal` для обертки модальных окон. В нем будет крестик, обработчик оверлея и `Escape`. Внутри него будет портал и `ModalOverlay` (это темное пространство вокруг попапа)



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, так как не исправлены все замечания прошлой итерации (тут https://github.com/vda-777/react-burger/pull/2#pullrequestreview-1378052929 ) :


* `style={{` используете 37 раз в коде проекта. Их все нужно убрать
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.
* `position: absolute` не стоит использовать в `css` в проекте. Это нигде не нужно (кроме крестика в попапе). Иначе невозможно сделать хоть какой-то минимальный адаптивный дизайн сайта. Все делается на `flex` или `grid` обычно, чтобы позиционировать правильно (и центрировать)
* и другие замечания прошлой итерации не исправлены https://github.com/vda-777/react-burger/pull/2#pullrequestreview-1378052929

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, 

* Вы прислали проект в проекте. Скрин https://disk.yandex.ru/i/G41FdvS-rTog6g . Вам нужно удалить лишние файлы и выделить только одни проект. Не понятно, что проверять
* Кое-что не исправили https://github.com/vda-777/react-burger/pull/2#discussion_r1162039734 
* И это не исправили https://github.com/vda-777/react-burger/pull/2#discussion_r1162040511 




**Можно лучше**


* `position: absolute` не стоит использовать в `css` в проекте. Это нигде не нужно (кроме крестика в попапе). Иначе невозможно сделать хоть какой-то минимальный адаптивный дизайн сайта. Все делается на `flex` или `grid` обычно, чтобы позиционировать правильно (и центрировать)

vda-777 committed:
2023-06-01

vda-777 committed:
2023-06-01 ver 2

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,53 @@
+              </div>
+            :
+              <div className={style.modalWindowHeaderCaption + ' mt-10'}>
+                    <p className="ml-10 text text_type_digits-medium">Детали ингредиента</p>
```
`Детали ингредиента`

Внутри компонента `Modal` не должно быть **хардкода** строк и компонентов. 



`Детали ингредиента` нужно передавать  в виде `header` или `title` в пропсы.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,53 @@
+  },[]);
+
+  return (    
+    <div className={style.modalOverlay}
```
По заданию должен быть компонент `ModalOverlay`, который используется внутри `Modal` в качестве оверлея

gennady-bars reviewed (commented):
Еще немного и все будет отлично:


* При открытии попапа ошибка в консоли https://disk.yandex.ru/i/CDP8TgwLKJ6MuQ 
* Внутри компонента `Modal` не должно быть **хардкода** строк и компонентов. 
* По заданию должен быть компонент `ModalOverlay`, который используется внутри `Modal` в качестве оверлея
* `ModalShow` не нужен в проекте: `createPortal` должен быть внутри `Modal`, чтобы не дублировать логику


**Можно лучше**

* Скрин https://disk.yandex.ru/i/BOuxqPeeoqmrtA  контент сайта не должен расплываться на больших экранах шире, чем 1280px. Весь контент сайта нужно поместить в `div` с классом `page`, которому нужно задать `margin: 0 auto` и `max-width: 1280px `, чтобы удерживать весь контент посередине экрана и не давать ему расширяться больше, чем положено.

vda-777 committed:
2023-06-02

При открытии попала ошибка в консоли - исправлено.
Компонент modal содержит createPortal + компонент ModalOverlay (весь хардкод в нем). ModalShow удвлен.
Контент сайта не должен расплываться на больших экранах шире, чем 1280px - доработано.

gennady-bars commented on src/components/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,53 @@
+        >
+          
+          {
+            props.header === undefined 
+            ?
+              <div className={style.modalWindowHeaderNoCaption}>
+                  <span className='mt-10 ml-10'><p className="text text_type_digits-medium">{props.header}</p></span>
+                  <span className='mt-15 mr-10'><CloseIcon type="primary" onClick={props.onClose}/></span>
+              </div>
+            :
+              <div className={style.modalWindowHeaderCaption + ' mt-10'}>
+                    <p className="ml-10 text text_type_digits-medium">Детали ингредиента</p>
```
`Детали ингредиента` нужно передавать в виде `title` или `header` в **пропсы** компоненту `Modal`. А `ModalOverlay` не должен знать про эту строку. Это тоже универсальный компонент темного пространства вокруг попапа


Вот так будет

```jsx

                    <p className="ml-10 text text_type_digits-medium">{props.header}</p>


```


Как раз у Вас это уже реализовано, но забыли заменить

gennady-bars replied:
Вам нужно вернуть весь код в `Modal`, а `ModaOverlay` должен содержать в себе только один блок (темное пространство вокруг попапа). И клик на этом пространстве, который закрывает попап

gennady-bars reviewed (commented):
Кое-что нужно подправить:


* Вам нужно вернуть весь код в `Modal`, а `ModaOverlay` должен содержать в себе только один блок (темное пространство вокруг попапа). И клик на этом пространстве, который закрывает попап
* `Детали ингредиента` нужно передавать в виде `title` или `header` в **пропсы** компоненту `Modal`. А `ModalOverlay` не должен знать про эту строку. Это тоже универсальный компонент темного пространства вокруг попапа

vda-777 committed:
2023-06-04

Вернул код в Modal, изменил ModalOverlay на один div передается один пропс на функцию закрытия, ловится кликом на нем и клавишей esc.
В Modal передается пропсы header - заголовок окна и children - разметка содержимого окна.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -2,15 +2,40 @@ import React from 'react';
+              </div>
+              :
+                <div className={style.modalWindowHeaderCaption + ' mt-10'}>
+                  <p className="ml-10 text text_type_digits-medium">Детали ингредиента</p>
```
`Детали ингредиента`  приходят в виде `props.header`. Не нужно хардкодить строки внутри `Modal`


Вы передаете уже все в пропсы, а тут должно быть вот так:

```jsx

                  <p className="ml-10 text text_type_digits-medium">{props.header}</p>


```

gennady-bars reviewed (commented):
Кое-что не исправили 


* `Детали ингредиента`  приходят в виде `props.header`. Не нужно хардкодить строки внутри `Modal`

vda-777 committed:
2023-06-05

Передается заголовок окна через props.header, содержание окна через props.children

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -15,7 +16,12 @@ export default function BurgerConstructor(props) {
+        <Modal onClose={() => setShowModal(false)} header={
+          <div className={style.modalWindowHeaderNoCaption}>
+            <span className='mt-10 ml-10'><p className="text text_type_digits-medium">{props.header}</p></span>
+            <span className='mt-15 mr-10'><CloseIcon type="primary" onClick={() => setShowModal(false)}/></span>
```
Суть компонента `Modal` в том, что он содержит в себе всю повторяющуюся верстку. Крестик всегда есть в попапе. Не нужно его передавать снаружи под каждый попап


Если Вы хотите немного изменять стили (когда есть заголовок, а когда его нет), то это можно делать через проверку, что передали `header`  в пропсы. 


Вот так тут будет

```jsx

      {showModal && (
        <Modal onClose={() => setShowModal(false)} header={""}>
          <OrderDetails numberOrder="034356" />
        </Modal>
      )}

```


header в пропсах будет пустой строчкой, значит, он будет `falsy`


Внутри `Modal` нужно будет проверять по 


```jsx
 {props.header? (
              // если есть хэдер, значит, рисуем его верстку и стили

// иначе рисует то, что для попапа заказа (без заголовка)

```

gennady-bars reviewed (commented):
К сожалению, работу придется отклонить без потери итерации


* Суть компонента `Modal` в том, что он содержит в себе всю повторяющуюся верстку. Крестик всегда есть в попапе. Не нужно его передавать снаружи под каждый попап

vda-777 committed:
2023-06-06

Сделал всю разметку под заголовок в modal без хардкода вроде. в hedaer передается только строка заголовка окна.

vda-777 committed:
2023-06-06 ver 2

убрал useEffect и useRef из modaloverlay и сделал в modal useEffect с подписку/отписку на события документа keydown

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -2,11 +2,18 @@ import React from 'react';
+  React.useEffect(() => {
+    document.addEventListener("keydown", (e)=>{if(e.key === 'Escape') props.onClose()});
+    return()=>{
+      document.removeEventListener("keydown", (e)=>{if(e.key === 'Escape') props.onClose()});
```
`      document.removeEventListener("keydown", (e)=>{if(e.key === 'Escape') props.onClose()});
`

В таком варианте невозможно удалить обработчик, так как Вы его **ссылку** не передаете. Там Вы передаете **анонимную** стрелочную функцию (у неё нет постоянной ссылки). Движок браузера не сможет найти по ссылке ее и удалить. Ошибки никакой браузер не выдаст, так как ему без разницы, что он не нашел функцию.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -2,11 +2,18 @@ import React from 'react';
-export default function Modal(props) {
+export default function Modal(props) {  
+  React.useEffect(() => {
+    document.addEventListener("keydown", (e)=>{if(e.key === 'Escape') props.onClose()});
```
Можно добавить обработчик `Escape` 

```jsx

  useEffect(() => {
    function closeByEscape(evt) {
      if(evt.key === 'Escape') {
        props.onClose();
      }
    }
      document.addEventListener('keydown', closeByEscape);
      return () => {
        document.removeEventListener('keydown', closeByEscape);
      }

  }, []) 

```

Объявляем функцию внутри `useEffect`, чтобы она не теряла свою ссылку при обновлении компонента.
И не забываем удалять обработчик в `clean up` функции через `return`
А также массив зависимостей

gennady-bars commented on src/components/modal-overlay/modal-overlay.jsx:
```diff
@@ -4,17 +4,17 @@ import PropTypes from 'prop-types';
+      //ref={refModalOverlay}
       onClick={props.onClose}
-      onKeyDown={(e)=>{if(e.key === 'Escape') props.onClose()}}
+      //onKeyDown={(e)=>{if(e.key === 'Escape') props.onClose()}}
```
Вам нужно очистить код от закомментированного кода. Он мешает  ревью

gennady-bars reviewed (commented):
Кое-что нужно подправить:

* Нужно правильно навешивать и удалять обработчик `Escape`
* Вам нужно очистить код от закомментированного кода. Он мешает  ревью

vda-777 committed:
07-06-2023

Добавил обработчик Escape в useEffect
Удалил закомментированный код.

vda-777 committed:
2023-06-07 ver 2

Удалили лишние <></> {} в createPortal

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

gennady-bars reviewed (commented):
*
